Source/NSString.m

```diff
@@ -6248,6 +6256,156 @@ - (BOOL) containsString: (NSString *)string
   return [self rangeOfString: string].location != NSNotFound;
 }
 
+- (void) enumerateSubstringsInRange: (NSRange)range 
+                            options: (NSStringEnumerationOptions)opts 
+                         usingBlock: (GSNSStringEnumerationBlock)block
+{
+  // Get low byte.
+  uint8_t substringType = opts & 0xFF;
+
+  BOOL isReverse = opts & NSStringEnumerationReverse;
+  BOOL substringNotRequired = opts & NSStringEnumerationSubstringNotRequired;
+  BOOL localized = opts & NSStringEnumerationLocalized;
+  
+  NSUInteger currentLocation;
+  BOOL stop = NO;
+  if(isReverse) {
+    currentLocation = range.location + range.length;
+  } else {
+    currentLocation = range.location;
+  }
+
+  if (substringType == NSStringEnumerationByLines || substringType == NSStringEnumerationByParagraphs) {
```

Comment:
I'm not actually sure. I wrote this code more than a year ago, but from my rereading it appears that it should use bitwise operations. Should I add another commit that fixes this issue, or is there a way to edit the current commit?

Reply:
This is actually a bit of a wierd enum. The lower byte is just a normal byte, with the possible values 0, 1, 2, 3, 4, 5, and 6 (5 and 6 are undocumented public APIs, so I didn't implement them). The second-lowest byte is a bitfield. Therefore, we have to get the lowest byte and we've stored it in `substringType`. Therefore, I think my implementation is correct.

Reply:
See the declaration:

```objc
enum {
  NSStringEncodingConversionAllowLossy = 1,
  NSStringEncodingConversionExternalRepresentation = 2
};
typedef NSUInteger NSStringEncodingConversionOptions;
#endif

#if OS_API_VERSION(MAC_OS_X_VERSION_10_6,GS_API_LATEST) 
/** For enumerateSubstringsInRange:options:usingBlock: 
    You must include an substring type (`NSStringEnumerationBy`), and may
    bitwise or (`|`) with any of the other options. */
enum {
    /* Must include one of these 
       Must fit into 8 bits. */
    /** Enumerate by lines. Uses lineRangeForRange: */
    NSStringEnumerationByLines = 0,
    /** Enumerate by paragraph. Uses paragraphRangeForRange: */
    NSStringEnumerationByParagraphs = 1,
    /** Enumerate by composed character sequence. Uses rangeOfComposedCharacterSequencesForRange: */
    NSStringEnumerationByComposedCharacterSequences = 2,
    /** Enumerate by word, as specified in Unicode TR 29. 
        Only supported if GNUstep is compiled with ICU. 
        Uses UBRK_WORD, with current locale and standard abbreviation lists if 
        NSStringEnumerationLocalized is passed, otherwise the locale is "en_US_POSIX". */
    NSStringEnumerationByWords = 3,
    /** Enumerate by sentence, as specified in Unicode TR 29. 
        Only supported if GNUstep is compiled with ICU. 
        Uses UBRK_WORD, with current locale and standard abbreviation lists if 
        NSStringEnumerationLocalized is passed, otherwise the locale is "en_US_POSIX". */
    NSStringEnumerationBySentences = 4,
    #if OS_API_VERSION(MAC_OS_X_VERSION_11,GS_API_LATEST) 
    /** Undocumented public API on macOS. Not supported by GNUstep. */
    NSStringEnumerationByCaretPositions = 5,
    /** Undocumented public API on macOS. Not supported by GNUstep. */
    NSStringEnumerationByDeletionClusters = 6,
    #endif

    /* May pass one of these via bitwise or. 
       Must be a single bit set at an offset >= 8. */
    NSStringEnumerationReverse = 1UL << 8,
    NSStringEnumerationSubstringNotRequired = 1UL << 9,
    NSStringEnumerationLocalized = 1UL << 10
};

typedef NSUInteger NSStringEnumerationOptions;
```

Reply:
See also how swift-corelibs-foundation does it: https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/NSString.swift#L827

They create a new enum that only contains the ...by... parts, and assume NSStringEnumerationByLines as default. I don't think that doing that is as correct, as Apple could add more and they would be considered ByLines instead of an unknown value.

Reply:
@fredkiefer Do you believe that my current approach is correct?

Reply:
What Apple does here is very confusing. This really isn#t good interface design. But we will have to follow and your interpretation sounds like the most sensible approach. When I wrote my comment I didn't at the actual documentation and just tried to make sense of the definition itself.